### PR TITLE
Connect command de/serialization 

### PR DESCRIPTION
**Pull request by YannickWeber**

### CODE REVIEW

File: src/main/java/org/varroa/load/MqttSlave.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.varroa.load;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.Future;
+
+public interface MqttSlave {
```
Comment (sbaier1) [inline src/main/java/org/varroa/load/MqttSlave.java]: wording: is this what we defined in the wiktionary as an MqttAgent? if yes, then it should be renamed accordingly.

File: src/main/java/org/varroa/chunk/command/ConnectCommand.java [hunk 2]
```diff
@@ -1,38 +1,34 @@
...
-    private final @NotNull MqttVersion mqttVersion;
-
-    private final @NotNull String id;
+    private final int port;
 
     public ConnectCommand(
-            @NotNull final MetaData metaData, @NotNull final String address, @NotNull final MqttVersion mqttVersion,
-            @NotNull final String id) {
-        super(metaData);
+            final @NotNull String brokerId, final @NotNull String address, final int port) {
```
Comment (sbaier1) [inline src/main/java/org/varroa/chunk/command/ConnectCommand.java]: why are we removing the mqtt version here? where is the protocol version determined by the executing client?

File: src/main/java/org/varroa/load/MqttContext.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package org.varroa.load;
+
+public interface MqttContext {
```
Comment (sbaier1) [inline src/main/java/org/varroa/load/MqttContext.java]: what is the point of this interface? javadoc please

File: src/main/java/org/varroa/chunk/command/Command.java [hunk 4]
```diff
@@ -1,20 +1,9 @@
...
 
-import org.jetbrains.annotations.NotNull;
-
 /**
  * @author Robin Atherton
  * @author Silvio Giebl
  */
 public abstract class Command {
 
-    private final @NotNull MetaData metaData;
```
Comment (sbaier1) [inline src/main/java/org/varroa/chunk/command/Command.java]: why are we removing the metadata field? is it not possible to serialize? we will probably need it at some point

Comment (SgtSilvio) [reply src/main/java/org/varroa/chunk/command/Command.java]: It is not needed for now. When it is needed we can add it back again.
It should not hold us back now.

File: src/main/java/org/varroa/load/Mqtt5Slave.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package org.varroa.load;
+
+import org.jetbrains.annotations.NotNull;
+import org.mqttbee.mqtt.MqttClient;
+import org.mqttbee.mqtt.MqttClientBuilder;
+
+import java.util.concurrent.Future;
+
+public class Mqtt5Slave implements MqttSlave {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/load/Mqtt5Slave.java]: This should not be part of this PR

File: src/main/java/org/varroa/remote/codec/decoder/command/ConnectDecoder.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import io.netty.buffer.ByteBuf;
+import org.jetbrains.annotations.NotNull;
+import org.varroa.chunk.command.ConnectCommand;
+
+import java.nio.charset.StandardCharsets;
+
+public class ConnectDecoder {
+
+    @NotNull ConnectCommand decode(final @NotNull ByteBuf in) {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/remote/codec/decoder/command/ConnectDecoder.java]: minor: imho this should be named `ConnectCommandDecoder` as it decodes a `ConnectCommand` not a `Connect`

File: src/main/java/org/varroa/remote/codec/decoder/command/CommandDecoders.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package org.varroa.remote.codec.decoder.command;
+
+import org.jetbrains.annotations.NotNull;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+public class CommandDecoders {
+
+    private final @NotNull Provider<ConnectDecoder> connectDecoderProvider;
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/remote/codec/decoder/command/CommandDecoders.java]: Why is the provider needed? The `ConnectDecoder` will be used anyway.

File: src/main/java/org/varroa/remote/codec/encoder/command/CommandEncoders.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+@Singleton
+public class CommandEncoders {
+
+    private final @NotNull Provider<ConnectEncoder> connectEncoderProvider;
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/remote/codec/encoder/command/CommandEncoders.java]: Is the Provider necessary? The `ConnectEncoder` will be used anyway.

File: src/main/java/org/varroa/remote/codec/encoder/command/ConnectEncoder.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import io.netty.buffer.ByteBuf;
+import org.jetbrains.annotations.NotNull;
+import org.varroa.chunk.command.ConnectCommand;
+import org.varroa.remote.message.ChunkMessage;
+
+import javax.inject.Singleton;
+import java.nio.charset.StandardCharsets;
+
+@Singleton
+public class ConnectEncoder {
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/remote/codec/encoder/command/ConnectEncoder.java]: minor: should be named `ConnectCommandEncoder`

File: src/main/java/org/varroa/load/MqttContext.java (hunk 3 above)
Comment (SgtSilvio) [reply src/main/java/org/varroa/load/MqttContext.java]: should not be part of this PR

File: src/main/java/org/varroa/load/MqttSlave.java (hunk 1 above)
Comment (SgtSilvio) [reply src/main/java/org/varroa/load/MqttSlave.java]: should not be part of this PR

File: src/main/java/org/varroa/remote/codec/decoder/ChunkMessageDecoder.java [hunk 10]
```diff
@@ -57,16 +64,15 @@
     private @NotNull Command decodeCommand(final @NotNull ByteBuf in) {
         switch (ChunkMessage.CommandType.values()[in.readByte()]) {
             case CONNECT:
-                return new ConnectCommand(new MetaData(), "dashboard.mqtt.com", MqttVersion.MQTT_3_1_1,
-                        NamePatternInstantiator.DEFAULT_PATTERN);
+                return new ConnectCommand("bk1", "dashboard.mqtt.com", 1883);
```
Comment (SgtSilvio) [inline src/main/java/org/varroa/remote/codec/decoder/ChunkMessageDecoder.java]: The `ConnectDecoder` must be used here

File: src/main/java/org/varroa/chunk/command/ConnectCommand.java (hunk 2 above)
Comment (s-gheldd) [reply src/main/java/org/varroa/chunk/command/ConnectCommand.java]: This is part of the client-group and not of the lifecycle. So I guess in the Chunk?

File: src/main/java/org/varroa/remote/codec/decoder/command/CommandDecoders.java (hunk 7 above)
Comment (s-gheldd) [reply src/main/java/org/varroa/remote/codec/decoder/command/CommandDecoders.java]: Idea was to enable caching of string serialization. As it is not implemented (yet) the Provider<> can go.

File: src/main/java/org/varroa/remote/codec/encoder/command/CommandEncoders.java (hunk 8 above)
Comment (s-gheldd) [reply src/main/java/org/varroa/remote/codec/encoder/command/CommandEncoders.java]: See comment above

**Comment by YannickWeber** [comment]
Comments were resolved in PR#36
